File: agents/research-runner/anthropic/runner.py
#!/usr/bin/env python3
"""Anthropic asyncio research runner.

Runs all 9 query templates in parallel against a single research question,
then writes a brief to the vault's wiki/briefs/ directory.

Usage:
    python runner.py "your research question"
    python runner.py --templates A1,M1 "targeted question"
    ANTHROPIC_API_KEY=sk-... python runner.py "question"

When ANTHROPIC_API_KEY is not set, falls back to `claude -p` subprocess
(works with a Claude Code subscription — no separate API key needed).
"""
from __future__ import annotations

import argparse
import asyncio
import json
import os
import subprocess
import sys
from pathlib import Path

# Allow importing from ../shared without installing as a package
sys.path.insert(0, str(Path(__file__).parent.parent / "shared"))

import anthropic
from dotenv import load_dotenv
from output import TemplateResult, make_async_anthropic_client, synthesize, write_brief
from templates import QueryTemplate, ResearchData, load_templates

load_dotenv()


async def run_template(
    client: anthropic.AsyncAnthropic,
    system_prompt: str,
    template: QueryTemplate,
    query: str,
) -> TemplateResult:
    """Send one template + query to Claude via SDK (requires ANTHROPIC_API_KEY)."""
    response = await client.messages.create(
        model="claude-opus-4-7",
        max_tokens=4096,
        # Cache the system prompt — identical across all 9 parallel calls,
        # so calls 2-9 hit the prompt cache at ~10% of full token cost.
        system=[
            {
                "type": "text",
                "text": system_prompt,
                "cache_control": {"type": "ephemeral"},
            }
        ],
        messages=[
            {
                "role": "user",
                "content": f"{template.text}\n\nQuery context: {query}",
            }
        ],
    )
    text = next(
        (block.text for block in response.content if block.type == "text"),
        "(no text response)",
    )
    return TemplateResult(id=template.id, name=template.name, text=text)


async def run_template_cli(
    system_prompt: str,
    template: QueryTemplate,
    query: str,
) -> TemplateResult:
    """Subprocess fallback: uses `claude -p` (works with Claude Code subscription)."""
    full_prompt = f"{system_prompt}\n\n{template.text}\n\nQuery context: {query}"

    def _run() -> subprocess.CompletedProcess[str]:
        return subprocess.run(
            [
                "claude", "-p",
                "--output-format", "json",
                "--model", "claude-opus-4-7",
                "--no-session-persistence",
                "--tools", "",
            ],
            input=full_prompt,
            capture_output=True,
            text=True,
            timeout=120,
        )

    result = await asyncio.to_thread(_run)
    if result.returncode != 0:
        raise RuntimeError(
            f"claude CLI error for {template.id}: {result.stderr.strip()}"
        )
    text = json.loads(result.stdout).get("result", "(no text response)")
    return TemplateResult(id=template.id, name=template.name, text=text)


async def main() -> None:
    parser = argparse.ArgumentParser(
        description="Run research templates in parallel via Anthropic API or claude CLI"
    )
    parser.add_argument("query", nargs="+", help="Research question")
    parser.add_argument(
        "--templates",
        default="",
        metavar="IDs",
        help="Comma-separated template IDs to run (default: all). E.g. --templates A1,M2",
    )
    args = parser.parse_args()
    query = " ".join(args.query)

    data: ResearchData = load_templates()

    if args.templates:
        filter_ids = {t.strip() for t in args.templates.split(",")}
        templates = [t for t in data.templates if t.id in filter_ids]
        if not templates:
            print(f"No templates matched: {args.templates}", file=sys.stderr)
            sys.exit(1)
    else:
        templates = data.templates

    api_key = os.environ.get("ANTHROPIC_API_KEY", "").strip()

    if api_key:
        print(f"Running {len(templates)} template(s) via SDK…", file=sys.stderr)
        client = make_async_anthropic_client()
        tasks = [run_template(client, data.system_prompt, t, query) for t in templates]
    else:
        print(
            f"No ANTHROPIC_API_KEY — running {len(templates)} template(s) via claude CLI…",
            file=sys.stderr,
        )
        tasks = [run_template_cli(data.system_prompt, t, query) for t in templates]

    results: list[TemplateResult] = list(await asyncio.gather(*tasks))

    synthesis = synthesize(query, results)
    brief_path = write_brief(query, results, variant="anthropic", synthesis=synthesis)

    print(f"Brief written to: {brief_path}", file=sys.stderr)


if __name__ == "__main__":
    asyncio.run(main())

File: agents/research-runner/anthropic/runner.ts
/**
 * Anthropic Claude Agent SDK research runner.
 *
 * Runs all 9 query templates in parallel against a single research question,
 * then writes a brief to the vault's wiki/briefs/ directory.
 *
 * Usage:
 *   npx tsx runner.ts "your research question"
 *   npx tsx runner.ts --templates A1,M1 "targeted question"
 *
 * Requires ANTHROPIC_API_KEY set in the environment (or .env at repo root).
 */

import { query } from "@anthropic-ai/claude-agent-sdk";
import { loadTemplates, type QueryTemplate } from "../shared/templates.ts";
import {
  synthesize,
  writeBrief,
  type TemplateResult,
} from "../shared/output.ts";

async function runTemplate(
  systemPrompt: string,
  template: QueryTemplate,
  research: string,
  agent: string | undefined,
): Promise<TemplateResult> {
  const agentPrompt = agent ? `\n\nAgent instruction: ${agent}` : "";
  const prompt = `${systemPrompt}${agentPrompt}\n\n${template.text}\n\nQuery context: ${research}`;
  let result = "(no text response)";

  for await (const message of query({
    prompt,
    options: {
      // No file-system or web tools — pure text synthesis.
      // An empty allowedTools list keeps the loop to a single turn.
      allowedTools: [],
    },
  })) {
    if ("result" in message && typeof message.result === "string") {
      result = message.result;
    }
  }

  return { id: template.id, name: template.name, text: result };
}

function parseArgs(): { query: string; templateFilter: string[]; agent?: string } {
  const args = process.argv.slice(2);
  let templateFilter: string[] = [];
  let agent = process.env.AI_ENG_AGENT?.trim() || undefined;
  const rest: string[] = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--templates") {
      templateFilter = (args[++i] ?? "").split(",").filter(Boolean);
    } else if (args[i] === "--agent") {
      agent = args[++i]?.trim() || agent;
    } else {
      rest.push(args[i]);
    }
  }

  const q = rest.join(" ").trim();
  return { query: q, templateFilter, agent };
}

async function main(): Promise<void> {
  const { query: research, templateFilter, agent } = parseArgs();
  if (!research) {
    console.error(
      'Usage: npx tsx runner.ts [--templates A1,M2] "research question"',
    );
    process.exit(1);
  }

  const data = loadTemplates();
  const templates =
    templateFilter.length > 0
      ? data.templates.filter((t) => templateFilter.includes(t.id))
      : data.templates;

  if (templates.length === 0) {
    console.error(`No templates matched: ${templateFilter.join(",")}`);
    process.exit(1);
  }

  console.error(`Running ${templates.length} template(s) in parallel via Claude Agent SDK…`);

  const results = await Promise.all(
    templates.map((t) => runTemplate(data.systemPrompt, t, research, agent)),
  );

  const synthesis = await synthesize(research, results);
  const briefPath = writeBrief(research, results, "anthropic", synthesis);
  console.error(`Brief written to: ${briefPath}`);
}

main().catch((err) => {
  console.error(err instanceof Error ? err.message : err);
  process.exit(1);
});
